src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java

```diff
@@ -149,4 +153,96 @@ void buildProductKeysQueryPredicate_WithSomeBlankProductKeys_ShouldBuildCorrectQ
         final QueryPredicate<Product> queryPredicate = service.buildProductKeysQueryPredicate(productKeys);
         assertThat(queryPredicate.toSphereQuery()).isEqualTo("key in (\"key1\", \"key2\")");
     }
+
+    @Test
+    void fetchMatchingProductsByKeys_WithBadGateWayExceptionAlways_ShouldFail() {
+        final FakeProductClient fakeProductClient = new FakeProductClient(new BadGatewayException());
+
+        initMockService(fakeProductClient);
+
+        final Set<String> keys =  new HashSet<>();
+        keys.add("productKey1");
+
+        CompletionStage<Set<Product>> future = service.fetchMatchingProductsByKeys(keys);
+
+        assertThat(errorExceptions).isEmpty();
+        assertThat(errorMessages).isEmpty();
+        assertThat(future.toCompletableFuture())
+                .failsWithin(60, TimeUnit.SECONDS)
+                .withThrowableOfType(ExecutionException.class)
+                .withCauseExactlyInstanceOf(BadGatewayException.class);
+    }
+
+    @Test
+    void fetchProduct_WithBadGatewayException_ShouldFail() {
+        final FakeProductClient fakeProductClient = new FakeProductClient(new BadGatewayException());
+
+        initMockService(fakeProductClient);
+
+        CompletionStage<Optional<Product>> future = service.fetchProduct("productKey1");
+
+        assertThat(errorExceptions).isEmpty();
+        assertThat(errorMessages).isEmpty();
+        assertThat(future.toCompletableFuture())
+                .failsWithin(60, TimeUnit.SECONDS)
+                .withThrowableOfType(ExecutionException.class)
+                .withCauseExactlyInstanceOf(BadGatewayException.class);
+
+    }
+
+    @Nonnull
```

Comment:
`Nonnull` annotation is unncessary.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/622/commits/4b68a84f3fd74b502354698b9d4005df30faf73c